base/float.jl

```diff
@@ -199,6 +199,17 @@ promote_rule(::Type{Float64}, ::Type{Float32}) = Float64
 widen(::Type{Float16}) = Float32
 widen(::Type{Float32}) = Float64
 
+promote_op{T<:Union{Float32,Float64}}(::typeof(trunc), ::Type{Signed}, ::Type{T}) = Int
```

Comment:
I don't think you need to do this for numbers. A generic definition like this seems to be enough:
```julia
promote_op{R<:Number,S<:Number}(op, ::Type{R}, ::Type{S}) = typeof(op(one(R), one(S)))
```

I'm currently working on it for PR https://github.com/JuliaLang/julia/pull/16996.